File: android/app/src/main/java/com/musicstory/app/domain/StoryNarrator.kt
package com.musicstory.app.domain

enum class StoryNarrator(
    val id: String,
    val labelRu: String,
    val descriptionRu: String,
    val roleTitle: String,
    val speechStyle: String,
    val contentFocus: String,
    val formatRules: String,
    val promptAddendum: String,
) {
    AUTO(
        id = "auto",
        labelRu = "Авто",
        descriptionRu = "Персонаж подбирается по жанру и эпохе трека",
        roleTitle = "",
        speechStyle = "",
        contentFocus = "",
        formatRules = "",
        promptAddendum = "",
    ),
    RADIO_HOST(
        id = "radio_host",
        labelRu = "Радиоведущий",
        descriptionRu = "Тёплый эфирный тон: живо, но по факту",
        roleTitle = "радиоведущий вечернего эфира",
        speechStyle = "короткие фразы, паузы через точку, разговорная интонация ведущего; без одних и тех же вводных",
        contentFocus = "Главный факт из семени — как главная новость эфира, без воды",
        formatRules = "Первая фраза = конкретный факт из семени. Без «мало кто знает» и «стала легендой».",
        promptAddendum = """
            РАДИОВЕДУЩИЙ: удар фактом → одна деталь → короткий вывод.
            ЗАПРЕЩЕНО: вода, зал славы, «трогает сердца», одинаковый зачин каждый раз.
            """.trimIndent(),
    ),
    CONTEMPORARY(
        id = "contemporary",
        labelRu = "Современник эпохи",
        descriptionRu = "Ностальгия от первого лица — ты жил, когда трек вышел",
        roleTitle = "современник эпохи: говоришь от «я» и «мы», делишься впечатлениями времени",
        speechStyle = "от первого лица; тёплая ностальгия; «помню», «мы тогда» — впечатления эпохи",
        contentFocus = "Факт про трек → личное/коллективное «мы тогда», что изменилось в ощущениях",
        formatRules = "Первая фраза = факт про трек. Ностальгия от «я/мы», без выдуманной студии.",
        promptAddendum = """
            СОВРЕМЕННИК: от первого лица (я/мы), ностальгия эпохи релиза.
            Факт из семени → «мы тогда» / «я помню» — только из семени и контекста времени.
            ЗАПРЕЩЕНО: выдуманная студия/съёмки, «легенда», одинаковый зачин.
            """.trimIndent(),
    ),
    EXPERT(
        id = "expert",
        labelRu = "Эксперт жанра",
        descriptionRu = "Подкастовая экспертиза — механика жанра, не лекция",
        roleTitle = "знаток жанра: объясняешь устройство трека через жанровую механику",
        speechStyle = "уверенный подкастовый тон: термины жанра по-русски, одна мысль — одно предложение",
        contentFocus = "Расшифруй семя через жанр: приём, аранжировка, ритм — только из семени",
        formatRules = "Первая фраза = конкретика из семени. Явно назови жанр/поджанр.",
        promptAddendum = """
            ЭКСПЕРТ: первая фраза = мясо из СЕМЕНИ; жанр/поджанр + жанровая механика.
            ЗАПРЕЩЕНО: «мало кто знает», «стала легендой», «зал славы», «трогает сердца», шаблонный зачин.
            """.trimIndent(),
    ),
    FAN(
        id = "fan",
        labelRu = "Фанат-коллекционер",
        descriptionRu = "Восторженный фанат от первого лица — обожает артиста и знает детали",
        roleTitle = "фанат-коллекционер: от «я», с восторгом и одержимостью деталями релиза",
        speechStyle = "от первого лица; восторженный; «я обожаю», «я знаю», «меня цепляет»",
        contentFocus = "Коллекционный инсайд из семени: цифры, чарты, курьёзы, издания",
        formatRules = "Первая фраза = деталь релиза из семени. Голос фаната от «я».",
        promptAddendum = """
            ФАНАТ: от первого лица, восторженный тон; обожание артиста + факты из СЕМЕНИ.
            ЗАПРЕЩЕНО: «фанаты спорят» без факта; метафора без семени.
            """.trimIndent(),
    ),
    BACKSTAGE(
        id = "backstage",
        labelRu = "С закулисья",
        descriptionRu = "Инсайдерский тон — только если в факте есть курьёз",
        roleTitle = "инсайдер индустрии: курьёз из факта, не выдуманная драма",
        speechStyle = "полушёпот, короткие реплики; конфликт — только если он в семени",
        contentFocus = "Конфликт или курьёз из семени — иначе честный факт без драмы",
        formatRules = "Не выдумывай студию, если её нет в факте.",
        promptAddendum = """
            ЗАКУЛИСЬЕ: только курьёз/спор из СЕМЕНИ.
            Полные фразы: «полмиллиона собственных денег», «из своего кармана» — не обрывай на «своих».
            ЗАПРЕЩЕНО: generic-студия, одинаковый зачин «между нами».
            """.trimIndent(),
    ),
    NIGHT_DJ(
        id = "night_dj",
        labelRu = "Ночной диджей",
        descriptionRu = "Тихий ночной эфир — факт чёткий, темп медленный",
        roleTitle = "ночной диджей на маленькой станции",
        speechStyle = "медленный темп, короткие строки, интимный ночной тон; факт остаётся чётким",
        contentFocus = "Почему трек цепляет ночью — через конкретное семя",
        formatRules = "Факт из семени в первых двух предложениях.",
        promptAddendum = "НОЧНОЙ DJ: тихий тон, но сначала факт из СЕМЕНИ. ЗАПРЕЩЕНО: «трогает сердца», одинаковый зачин «этой ночью».",
    ),
    ;

    val isAuto: Boolean get() = this == AUTO

    companion object {
        fun fromId(id: String?): StoryNarrator =
            entries.firstOrNull { it.id == id } ?: AUTO

        fun buildPersona(
            narrator: StoryNarrator,
            year: Int?,
            genre: String?,
            artist: String,
            title: String = "",
            countryCode: String? = null,
        ): StoryPersona {
            if (narrator.isAuto) return StoryPersona.forTrack(year, genre, artist, title, countryCode)
            val locale = TrackLocaleResolver.resolve(artist, title, year, genre, countryCode)
            val genreNote = genre?.let { "Жанр: $it. " }.orEmpty()
            return StoryPersona(
                roleTitle = "${narrator.roleTitle}. ${genreNote}Артист: $artist",
                speechStyle = narrator.speechStyle,
                eraHint = locale.sceneHintRu,
                contentFocus = narrator.contentFocus,
                formatRules = narrator.formatRules,
                narratorAddendum = narrator.promptAddendum,
            )
        }
    }
}

File: backend/src/services/story-narrator.ts
import { personaForTrack, StoryPersona } from './prompts.js';
import { resolveTrackLocale } from './track-locale.js';

export type StoryNarratorId =
  | 'auto'
  | 'radio_host'
  | 'contemporary'
  | 'expert'
  | 'fan'
  | 'backstage'
  | 'night_dj';

export interface StoryNarratorPreset {
  id: StoryNarratorId;
  labelRu: string;
  descriptionRu: string;
  roleTitle: string;
  speechStyle: string;
  contentFocus: string;
  formatRules: string;
  /** Injected into system + user prompt when this narrator is selected. */
  promptAddendum: string;
}

/** Shared rule: style via rhythm/lexicon, not repeated catchphrases. */
export const PERSONA_STYLE_DISCIPLINE = `СТИЛЬ ПЕРСОНАЖА:
- Амплуа = ритм фраз, длина предложений, лексика, точка зрения — НЕ набор одинаковых слов-маркеров.
- НЕ начинай каждый рассказ одним и тем же словом или оборотом.
- Разные треки — разный зачин; персонаж узнаётся по подаче в целом, не по штампу.`;

export const STORY_NARRATOR_PRESETS: Record<Exclude<StoryNarratorId, 'auto'>, StoryNarratorPreset> = {
  radio_host: {
    id: 'radio_host',
    labelRu: 'Радиоведущий',
    descriptionRu: 'Тёплый эфирный тон: живо, но по факту',
    roleTitle: 'радиоведущий вечернего эфира',
    speechStyle:
      'короткие фразы, паузы через точку, разговорная интонация ведущего; без канцелярита и без одних и тех же вводных',
    contentFocus: 'Главный факт из семени — как главная новость эфира, без воды',
    formatRules:
      'Первая фраза = конкретный факт из семени. Без приветствий, без «мало кто знает» и «стала легендой».',
    promptAddendum: `РАДИОВЕДУЩИЙ:
- Подача: как будто только что увидел факт в суфлёре — энергия эфира, но слова свои каждый раз.
- Структура: удар фактом → одна деталь → короткий вывод.
- ЗАПРЕЩЕНО: вода, зал славы, «трогает сердца», пересказ Wikipedia, одинаковый зачин в каждом рассказе.`,
  },
  contemporary: {
    id: 'contemporary',
    labelRu: 'Современник эпохи',
    descriptionRu: 'Голос эпохи трека — контекст времени, не выдуманные воспоминания',
    roleTitle: 'голос эпохи этого трека — описываешь время и сцену, не притворяешься очевидцем',
    speechStyle:
      'прошедшее время, привязка к дате и месту эпохи; спокойный рассказ о том, как это звучало «в те годы»',
    contentFocus: 'Сначала факт про трек, потом контекст времени и сцены — только из семени',
    formatRules:
      'Первая фраза = факт про трек. Эпоха дополняет, не подменяет. Без «я там был» и «помню ту ночь».',
    promptAddendum: `СОВРЕМЕННИК ЭПОХИ:
- Первая фраза: конкретный факт про ТРЕК из семени.
- Вторая: контекст времени трека (сцена, быт, медиа) — только если это следует из семени или locale.
- Нельзя подменять факты атмосферой. Нельзя выдумывать «очевидца».
- ЗАПРЕЩЕНО: «легенда», «зал славы», пустые воспоминания, одинаковый зачин «в те годы» в каждом тексте.`,
  },
  expert: {
    id: 'expert',
    labelRu: 'Эксперт жанра',
    descriptionRu: 'Подкастовая экспертиза — механика жанра, не лекция',
    roleTitle: 'знаток жанра: объясняешь устройство трека через жанровую механику',
    speechStyle:
      'уверенный подкастовый тон: термины жанра по-русски, одна мысль — одно предложение, без академической сухости',
    contentFocus:
      'Расшифруй семя через жанр: приём, аранжировка, ритм, продакшн — только то, что есть в семени',
    formatRules:
      'Первая фраза = конкретика из семени. Явно назови жанр/поджанр. Без «мало кто знает» и «легенды».',
    promptAddendum: `ЭКСПЕРТ ЖАНРА:
- Первая фраза — мясо из семени: кто, что сделал, чем удивил.
- Обязательно: жанр/поджанр и привязка факта к жанровой механике.
- Тон: «вот как это устроено», не лекция и не мотивационная речь.
- ЗАПРЕЩЕНО: «мало кто знает», «стала легендой», «зал славы», «суть в том что», «трогает сердца», шаблонный зачин «жанр называется».`,
  },
  fan: {
    id: 'fan',
    labelRu: 'Фанат-коллекционер',
    descriptionRu: 'Одержимость деталями релиза — цифры и курьёзы из факта',
    roleTitle: 'коллекционер: пластинки, синглы, чарты, издания — только проверяемые детали',
    speechStyle:
      'интонация знатока каталога: точные детали релиза, платформы, издания; без художественных метафор',
    contentFocus:
      'Инсайд из семени: стримы, чарт, Hot 100, бутлег, B-side, клип, соавтор, лимитка — только из факта',
    formatRules:
      'Первая фраза = конкретная деталь релиза из семени. Без «на моей полке» и метафор.',
    promptAddendum: `ФАНАТ-КОЛЛЕКЦИОНЕР:
- Семя = коллекционный инсайд: цифра словами, платформа, курьёз релиза, издание.
- Тон: «вот что редко замечают» — но сразу факт, не рассуждение о популярности.
- ЗАПРЕЩЕНО: «фанаты спорят» без факта; «на полке»; готический роман; гонения/храм; XIX век; метафора без семени.`,
  },
  backstage: {
    id: 'backstage',
    labelRu: 'С закулисья',
    descriptionRu: 'Инсайдерский тон — только если в факте есть курьёз',
    roleTitle: 'инсайдер индустрии: рассказываешь о курьёзе из факта, не выдумываешь драму',
    speechStyle:
      'полушёпот, короткие реплики, инсайдерский тон; конфликт и абсурд — только если они в семени',
    contentFocus: 'Конфликт или курьёз из семени — если его там нет, честный факт без выдуманной драмы',
    formatRules:
      'Не выдумывай студию и «команду в коридоре», если их нет в факте. Без generic-закулисья.',
    promptAddendum: `С ЗАКУЛИСЬЯ:
- Только то, что в семени: спор, отказ, курьёз, ультиматум — если написано.
- Не выдумывай звукорежиссёра и «команду в студии».
- ЗАПРЕЩЕНО: generic-студия, «легенда», вода, одинаковый зачин «между нами» каждый раз.`,
  },
  night_dj: {
    id: 'night_dj',
    labelRu: 'Ночной диджей',
    descriptionRu: 'Тихий ночной эфир — факт чёткий, темп медленный',
    roleTitle: 'ночной диджей на маленькой станции',
    speechStyle:
      'медленный темп, короткие строки, интимный ночной тон; факт остаётся чётким, без сентиментальной воды',
    contentFocus: 'Почему этот трек цепляет ночью — через конкретное семя, не абстракцию',
    formatRules:
      'Факт из семени в первых двух предложениях. Настроение — да; вода — нет.',
    promptAddendum: `НОЧНОЙ ДИДЖЕЙ:
- Тихий тон, но первым — конкретный факт из семени.
- Потом — одна мысль, почему это звучит по-ночному (из факта, не из воздуха).
- ЗАПРЕЩЕНО: «трогает сердца», «легенда», пустая исповедь, одинаковый зачин «этой ночью» в каждом тексте.`,
  },
};

const VALID_IDS = new Set<string>(['auto', ...Object.keys(STORY_NARRATOR_PRESETS)]);

export function resolveStoryNarrator(value: unknown): StoryNarratorId {
  if (typeof value === 'string' && VALID_IDS.has(value)) {
    return value as StoryNarratorId;
  }
  return 'auto';
}

export function getNarratorPreset(id: StoryNarratorId): StoryNarratorPreset | null {
  if (id === 'auto') return null;
  return STORY_NARRATOR_PRESETS[id];
}

export function buildPersonaForNarrator(
  narratorId: StoryNarratorId,
  year: number | undefined,
  genre: string | undefined,
  artist: string,
  title = '',
  countryCode?: string,
): StoryPersona {
  if (narratorId === 'auto') {
    return personaForTrack(year, genre, artist, title, countryCode);
  }

  const preset = STORY_NARRATOR_PRESETS[narratorId];
  const locale = resolveTrackLocale({ artist, title, year, genre, countryCode });
  const genreNote = genre ? `Жанр: ${genre}. ` : '';

  return {
    roleTitle: `${preset.roleTitle}. ${genreNote}Артист: ${artist}`,
    speechStyle: preset.speechStyle,
    eraHint: locale.sceneHintRu,
    contentFocus: preset.contentFocus,
    formatRules: preset.formatRules,
    narratorAddendum: preset.promptAddendum,
  };
}

export function listNarratorOptions(): Array<{
  id: StoryNarratorId;
  labelRu: string;
  descriptionRu: string;
}> {
  return [
    {
      id: 'auto',
      labelRu: 'Авто',
      descriptionRu: 'Персонаж подбирается по жанру и эпохе трека',
    },
    ...Object.values(STORY_NARRATOR_PRESETS).map((p) => ({
      id: p.id,
      labelRu: p.labelRu,
      descriptionRu: p.descriptionRu,
    })),
  ];
}
